compiler/rustc_resolve/src/check_unused.rs

```diff
@@ -173,7 +173,7 @@ impl<'a, 'b, 'tcx> Visitor<'a> for UnusedImportCheckVisitor<'a, 'b, 'tcx> {
             self.base_use_tree = Some(use_tree);
         }
 
-        if self.base_use_is_pub {
+        if self.base_use_is_pub && !self.r.empty_glob_reexports.contains(&id) {
```

Comment:
```suggestion
        if self.r.effective_visibilities.is_exported(self.r.local_def_id(id)) {
```
We have effective visibilities available early now, so we not only don't need a separate `empty_glob_reexports` table here, but also can do much better than just checking `pub` in general.

Reply:
I suspect that there might be some bugs associated with the line ﻿`let vis = self.r.effective_visibilities.effective_vis(self.r.local_def_id(id));`.

for example:

```rs
mod bar {
    pub struct Bar;
}

fn f() {
    pub use bar::Bar; // the value of `vis` is `None`
}

pub use bar::Bar; 
//^ but the value of `vis` here is:
//  Some(
//    EffectiveVisibility {
//        direct: Public,
//        reexported: Public,
//        reachable: Public,
//        reachable_through_impl_trait: Public,
//    },
// )

```

Reply:
Both `vis` look correct in the snippet above.

Reply:
Fantastic, I will investigate it.

Reply:
Indeed, `﻿resolver.effective_visibilities.is_exported(local_def_id)` substantially aids us in detecting additional unused imports.